### ISSUE TITLE
GraphQL create item with relation should be checked against the update permissions not create

### ISSUE THREAD

**Issue by clck-dr** [directus/directus#7237]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

Related to following discussion: #7185

I want ot create an item, let's say activity with a related activitytype via GraphQL API.

GraphQL throws Validation Exception becaus of missing create_activitytype_input fields (e.g. name). The activitytype with id 1 has allready been created and the request should use update_activitytype_input instead (mentioned by rijkvanzanten).

### To Reproduce

POST /graphql
```
mutation {
    create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: {
              id: 1
            } 
        },
    )     
}
```

### What version of Directus are you using?

directus/directus:9.0.0-rc.85

### What version of Node.js are you using?

directus/directus:9.0.0-rc.85

### What database are you using?

mysql:5.7

### What browser are you using?

Version 92.0.4515.131 (Offizieller Build) (64-Bit)

### What operating system are you using?

Windows

### How are you deploying Directus?

Docker

**Comment by rijkvanzanten** [comment]
I'm wondering if this is something that we can actually do.. GraphQL doesn't allow for multiple types per input type field, so we can't have `activitytype` be "create or update". We might have to replace the `activitytype` field with  `create_activitytype` and `update_activitytype` fields 🤔 

thoughts @jaycammarano?

**Comment by jaycammarano** [comment]
My notes say:
We need to create a way to check if the user is trying to create or update a field. We could check for the existence of a PK(an optional GQL field). We would have to also check if the supplied PK is in the database, as the user can manually choose a PK on create.

Something to note: by not renaming the nested field, we will be implementing this change to the root of the mutation as well. This would allow for `articletype` to be called at root not just nested. I'm not sure about the ramifications of that. 

There may be a way to only allow `articletype` only when nested. I would have to look at the exact code for the mutations generator but it could be similar to 
https://github.com/directus/directus/blob/62051e462c6acedcb8796213a814a66d167b0e26/api/src/services/graphql.ts#L588-L644
Where the nests are made by first initializing the lowest fields and then attaching them to the resolver. In this issue, it would be attaching activitytype+activityFields to the create_activite_item. Then attaching those fields to the resolver as optional but not creating a separate resolver for those fields. This wouldn't make `articletype` its own resolver but rather just fields on the `create_ activity_item` resolver.

**Comment by 8byr0** [comment]
Are there any updates on this ? 

IMHO there are two ways to handle this problem:
1. Assume that all relations do not exist and need to be created on the go(actual behavior)
```graphql
mutation {
    create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: {
              id: 1
            } 
        },
    )     
}
```
2. Assume that your relations actually exist and you just pass ID of an existing item (strapi does this)
```graphql

**Comment by 8byr0** [comment]
Are there any updates on this ? 

IMHO there are two ways to handle this problem:
1. Assume that all relations may not exist and need to be created on the go(actual behavior) or updated during every single mutation (create or update)
```graphql
mutation {
    create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: {
              id: 1
            } 
        },
    )     
}
```
2. Assume that your relations actually exist before running the mutation and only passing ID of an existing item (for instance strapi does this)
```graphql
mutation {
    create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: 1 
        },
    )     
}
```


I think the second option is the best one for a simple reason: your mutation `create_activity_item` should only serve that purpose, same for `update_activity_item`: it's not `update_activity_item_and_all_items_related`. 

If it also updates other elements then you delegate some logic onto the client side which is really risky and not efficient (you need to have more data available in order to make your mutation, and you need it every time you want to update an item - this behavior is actually the same for `update_activity_item`).

My suggestion would be to use an `Int!` type for all relations instead of `create_<whatever>`.

An union could also be possible, don't you think so?
```graphql 
union create_or_update_activity = Int | create_activity_item
type create_activity_input {
    id: ID
    sort: Int
    activityType: create_or_update_activity
}
```

**Comment by rijkvanzanten** [comment]
> Assume that all relations may not exist and need to be created on the go(actual behavior) or updated during every single mutation (create or update)

I think I would actually prefer option 1! While you would have to nest the id column, it does allow for both operations to be done, without us being opinionated about what's the right thing to do. Not being able to nested updates would be very inconvenient, as you'd have to do each insertion as a separate API call, in order from "bottom to top" yourself.

> An union could also be possible, don't you think so?

I wish, GraphQL doesn't support union input types.

**Comment by 8byr0** [comment]
> I think I would actually prefer option 1! While you would have to nest the id column, it does allow for both operations to be done, without us being opinionated about what's the right thing to do. Not being able to nested updates would be very inconvenient, as you'd have to do each insertion as a separate API call, in order from "bottom to top" yourself.

I agree with that, being able to update in one query is awesome. But having all the nested items constraints is really a pain when all you want is to setup a relation between your item and the one being contained.

 
 
> I wish, GraphQL doesn't support union input types.

Ah damn, just realised... There's an ongoing RFC https://github.com/graphql/graphql-spec/pull/733 about this (not input union, another type for flexibility) but may not arrive before a while...

**Comment by rijkvanzanten** [comment]
> but having all the nested items constraints is really a pain when all you want is to setup a relation between your item and the one being contained.

Agreed! It's a matter of which of the two options is the _least bad_ I'm afraid. In this particular case, I think having the option to do the nested insert is more important than the ability to have the "Flat key" in terms of feature set, would you agree? 🤔 

Alternatively, we could add a new mutation altogether for `create_nested_activity_item()` to switch between the two, but that feels very overkill for such a small benefit

**Comment by 8byr0** [comment]
> Agreed! It's a matter of which of the two options is the _least bad_ I'm afraid. In this particular case, I think having the option to do the nested insert is more important than the ability to have the "Flat key" in terms of feature set, would you agree? 🤔

What about a data structure with an optional payload?
```graphql
mutation {
  create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: {
                id: 1
                # Wrap all fields under an optional payload value, set it when you want to update / create on the fly 
                payload: {
                   whateverKey: "whateverValue"
                } 
            }
        },
    )   
}

 
> Alternatively, we could add a new mutation altogether for `create_nested_activity_item()` to switch between the two, but that feels very overkill for such a small benefit

Sounds a bit too much to me too, would involve a lot of useless complexity

**Comment by rijkvanzanten** [comment]
> What about a data structure with an optional payload?

What benefit does that have over the regular nested item? The only thing it'd add is another level of nesting right?

The ultimate goal would be to allow the user to do both:

```graphql
mutation {
  create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: {
                id: 1
                whateverKey: "whateverValue"
            }
        },
    )   
}
```

OR

```graphql
mutation {
  create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: 1
        },
    )   
}
```

But seeing that isn't a thing in GQL, I think the first of the two is the better approach 🤔

**Comment by 8byr0** [comment]
> > What about a data structure with an optional payload?
> 
> What benefit does that have over the regular nested item? The only thing it'd add is another level of nesting right?

The main benefit of this approach is the fact that payload is optional, not its content. Basically it allows:
- specifying only id if you want to link to an existing entity (do not fill payload)
- specify id and payload to perform a creation (then payload type would be `create_<item>_input` with all the creation constraints 


> The ultimate goal would be to allow the user to do both:

Sure, no debate on that point: this is actually not possible. But I'm not sure that the first approach is the best one, here's a simple example: 
1. Statement 
- A User collection contains a relation to a `company` collection
- Company item has some fields mandatory (legal identifier, vat, taxIdentifier, legalNotice and address for example - but there may be many more in this particular case)
- A company already exists in my database

2. User creation
- I have my user data ready for the mutation (for example, out of a form)
- since my company already exists, all I want to do is to pass its id in the create_user mutation.
- Due to the chosen approach, I need to make a query before my user creation to fetch the company data (mandatory fields) in order to fill them in my input - fields that I may not need afterwards. 

3. **Result**: 
- 1 query (for some fields I basically do not need in my client scope) 
- 2 mutations (the nested one will actually update the company item)

Another downside: if a required field is added to the company afterwards and I forgot to update my query, it will most likely fail next time it runs (yeah, should have written tests for this one 🛂 ).


With the other approach (only id, no payload), you only have:
- 1 query (to fetch the company id)
- 1 mutation (that will only alter a single collection)


One last point: consider having multiple levels of relation you have to collect all required fields of each nested relation in order to make your mutation

Does it make sense to you or am I totally wrong on this ?

**Comment by rijkvanzanten** [comment]
> specify id and payload to perform a creation (then payload type would be create_<item>_input with all the creation constraints

Ah yes, the problem here is the difference between _update_ and _create_ permissions though, not the difference between _create_ permissions or no-permissions check at all 🤔 

It's again a problem of union typing in GraphQL, as the switch between create/update technically has to be based on whether or not an item with that ID exists, which is not a thing we can expose in the static schema in GraphQL..


Regarding your examples:

1 & 2: This should be fine from Directus' perspective, as the fact that you're using an existing ID would kick in the update permissions, not the create permissions. That being said, we have to expose a single type in GraphQL, which would mean we could expose all fields but make them all not-required, only show a intersection of the available fields from update and create, or simply make the whole nested input object a JSON blob, and have Directus figure it out from there (effectively working around GraphQL's limitations).

> Does it make sense to you or am I totally wrong on this ?

**Comment by 8byr0** [comment]
Checkout how graphcms handles nested mutations (create and upsert mostly interesting for our use case): 
https://graphcms.com/docs/api-reference/content-api/mutations#nested-mutations

It creates a few levels of nesting (relationField > create or update > payload data) but makes queries verbose and consistent

**Comment by rijkvanzanten** [comment]
hmm yeah, sort of necessary evil approach imo. that being said, it does solve for this particular graphql issue

**Comment by 8byr0** [comment]
As told earlier I'm quite new to directus and not familiar yet with the codebase so I cannot really figure out how costly it would be to change the way graphql logic handles this in the backend. 
I guess it would be interesting to collect opinion of core and long term maintainers and make them join the discussion !

**Comment by rijkvanzanten** [comment]
Not super costly per se, just an unfortunate breaking change. The thing that bothers me most is that it's a fairly "ugly" addition we'll have to make to the structure in GraphQL, just because GraphQL itself has limitations in corners you wouldn't expect 😅 (this isn't the first time we've had to make concessions on the graphql side just because of limitations in the spec)

**Comment by 8byr0** [comment]
Sure, developers don't like breaking changes... But on the other hand developers do not like useless lines of code neither!

I'm really disappointed about this particular GraphQL limitation which is quite annoying but if you want to provide an advanced API covering as much use cases as possible and also more robust and complete than others (who have to deal with this limitation too), I think it's worse it!

**Comment by rijkvanzanten** [comment]
cc @jaycammarano

**Comment by ravi-2912** [comment]
Reading this discussion, it seems it is preferred to have create for the related item when creating an M2O object. If this is the case, then a standard Post/Comment cannot be implemented in Directus which I think could be a huge drawback.

**Comment by ravi-2912** [comment]
I just checked that if using REST api, I can just provide the ID of the related field. This is inconsistent with GraphQL API. Could you please check and make amends to the GraphQL API.

**Comment by rijkvanzanten** [comment]
> I just checked that if using REST api, I can just provide the ID of the related field. This is inconsistent with GraphQL API. 

That's not going to be supported in GraphQL, as GraphQL doesn't support union input types 👍🏻

**Comment by rijkvanzanten** [comment]
> No what I was saying that, there is no requirement for Union types, just allow the id to be supplied in the GraphQL schema. That way it will be consistent with the REST.

No, cause then you'd lose the ability to provide a nested object to do the create/update, which will be inconsistent with REST too 🔁

**Comment by ravi-2912** [comment]
Oh ok. Yeah I jsut tried that. REST allows both. Fair enough.

**Comment by josias-r** [comment]
```
mutation test2 {
  people {
    create(data: $var) {
      id
      name
    }
  }
  people {
    edit(id: 123, data: $var) {
      id
      name
    }
  }
}
```

This approach (also for nested items) is not an option? 🤔

**Comment by rijkvanzanten** [comment]
@josias-r Technically it would be, though it would be a shame if we'd have to add a non-standard approach like that that's different from the other entry points (like rest/cli/sdk) 🤔

**Comment by rijkvanzanten** [comment]
> What's so non-standard about it?

It's non-standard, as there's no real spec around what fields exist within a graphql query. There's no rules / agreed upon standards when it comes to field names for mutations. `create` / `update` differentiation exists on the parent level to mirror post/patch/delete, whereas the nested levels work with upserts basis, just like the rest api does. It'd be unfortunate if we'd have to resolve to another level of abstraction in CRUD on the nested level, as that'd be "non-standard" in either graphql standards (as they don't really exist), or in consistency with rest 🤔

**Comment by josias-r** [comment]
But from my understanding currently the `upsert` doesn't work at all. The example from the ticket:

```
mutation {
    create_activity_item(
        data: {
            start: "2021-08-04T07:06:10Z"
            time: "00:01:11"
            calculatedpoints: 0
            activitytype: {
              id: 1
            } 
        },
    )     
}
```
...shouldn't fail if the API was following the upsert basics.

Because the upsert process would be this:
1. Check if exists activitytype with id: 1
2. upsert
   - If exists update that item + set relation -> finished
   - If it doesn't exist create a new item with that id + set relation -> finished

But currently it always `inserts` and does not try to update first

Sure, giving the fine-control to the dev might be a powerful/flexible api, but as you mentioned, this might be hard to implement. And in most cases following the upsert-basics is perfectly I think

.. So isn't the main problem currently, that we are not actually following the `upsert` strategy?

**Comment by rijkvanzanten** [comment]
@josias-r That's exactly what the OP is describing, and why this is an open bug Issue 😉

> .. So isn't the main problem currently, that we are not actually following the upsert strategy?

Yes!

**Comment by josias-r** [comment]
@rijkvanzanten Alright gotcha 😁
Then yeah, upsert implementation seems to be the way to go.

You might have to fetch the previous IDs of the relationships that existed, if you want to update them. But once you have the nested IDs you can pretty much do anything

- Set relationship by passing ID
- Set+update relationship by passing ID and at the same time update the data b passing additional values
- Create relationship by not passing ID
- Create relationship including data by not passing ID and additional data

**Comment by rijkvanzanten** [comment]
Yeah! The Directus services already work like that (and that's exactly how it currently works on the REST API), so the "only"[^1] update here is that we gotta make sure that GraphQL is typed accordingly. We should make a new "upsert input type" that handles those differences. That being said, GraphQL doesn't support union input types, so we can't do point 1 (`Set relationship by passing ID`) it _has_ to be `{ id: 1 }` instead of `1`, but that's OK.

[^1]: quotes because that's not as easy as it sounds

**Comment by rijkvanzanten** [comment]
Linear: [ENG-293](https://linear.app/directus/issue/ENG-293/graphql-create-item-with-relation-should-be-checked-against-the-update)

**Comment by anthlasserre** [comment]
Hey guys 👋🏼 Do we have any updates there ?

**Comment by rijkvanzanten** [comment]
> Hey guys 👋🏼 Do we have any updates there ?

@anthlasserre Unfortunately not. I was hoping this might change, but GraphQL doesn't support a union input type (still), so we'll have to come up with an alternative. I'm currently torn between going with a different naming convention, where you have to pass an explicit `nested_field_create` vs `nested_field_update` etc, or treat every nested change as an upsert where we check for create _or_ update permissions, and then handle the final permissions check in Directus' data pipeline rather than the GraphQL schema 🤔 Any thoughts on that?

**Comment by ryami333** [comment]
Could `Create_Foo_Input` have a "special" property called something like `_relationships`, that can be `String`-typed (rather than `Create_Bar_Input`-typed?

So usage would look like:

```graphql
mutation {
  create_post_item(
    data: {
       title: "How to peel a banana"
       # This is how one would do a nested create:
       category: {
         title: "Tutorials"
      }
      # This is how one would do an association with an existing Category:
       _relationships: {
         category: "a8f5f167f44f4964e6c998dee827110c"
       }  
    }
  )     
}
```

**Comment by rijkvanzanten** [comment]
@ryami333 we could! It's a bit of shame though that we'd need additional "fake" fields to get around a graphql limitation. We'd also have to add multiples (create/update/delete), and we might not be able to use `_` as a prefix (i'm a little hazy if GQL allows that)

**Comment by ryami333** [comment]
Alternatively, it could be _not_ fake fields, but rather a separate argument altogether to `data`. Eg:

```graphql
mutation {
  create_post_item(
    data: {
      title: "How to peel a banana"
      # This is how one would do a nested create:
      category: { title: "Tutorials" }
    }
    # This is how one would do an association with an existing Category:
    relationships: {
      category: "a8f5f167f44f4964e6c998dee827110c"
    }
  )
}
```

**Comment by ryami333** [comment]
Also, for the sake of reference and documentation, here are the upstream blocking GraphQL RFC/PRs:

- Failed RFC for union input types: https://github.com/graphql/graphql-spec/issues/488
- Open succeeding pull-request for `@oneOf` directive: https://github.com/graphql/graphql-spec/pull/825

I would encourage everyone to go and upvote [that PR](https://github.com/graphql/graphql-spec/pull/825) to help unblock Directus' maintainers here.

**Comment by rijkvanzanten** [comment]
Interesting suggestion! Would you prefer that over a slightly-less type safe "upsert" type[^1] that handles both underneath the data field? If so, why?

[^1]: My current _next best thing_ option is to auto-generate an upsert-type that effectively merges the create and update types, and have the exact validation of fields handled by the Directus engine rather than GraphQL typing, but that does go against GraphQL common usage a bit.

**Comment by ryami333** [comment]
I don't have a strong preference because either approach has tradeoffs, but I suppose the upsert type would be ergonomically closer to the `@oneOf` operator, so more future-proof. In _either_ case, the types would have to be loosened and validation would have to be deferred to the GraphQL engine, because you'd have to make non-nullable relationships nullable, otherwise this would fail:

```graphql
mutation {
  create_post_item(
    data: {
      title: "How to peel a banana"
      # GraphQL Validation error: category may not be null.
    }
    relationships: {
      category: "a8f5f167f44f4964e6c998dee827110c"
    }
  )
}
```

**Comment by rijkvanzanten** [comment]
Good points! I think I personally prefer going that "faked" `@oneOf` approach, as it's closer in line with rest/websocket/sdk usage 🙂 Here's to hoping that `@oneOf` syntax will become a reality at some point

**Comment by ryami333** [comment]
> Here's to hoping that @oneOf syntax will become a reality at some point

Good news @rijkvanzanten, `oneOf` was introduced to `graphql-js` in [release v16.9.0](https://github.com/graphql/graphql-js/releases/tag/v16.9.0).

**Comment by rijkvanzanten** [comment]
Awesome! It does require you to provide the field name of the object you're inputting, but that's a lot nicer for nested updates than having to rely on a less-strictly typed upsert approach. It'd be a massive breaking change, but we could consider having a single "Mutation field" now that then relies on the `oneOf` inputs, which then relies on more nested one-of input types for it's relational content. In a simple articles->author relationship 

```graphql
# ⚠️ This is pseudo-code, do not copy+paste ⚠️
input ArticlesInput @oneOf {
  create: ArticlesInputCreate
  update: ArticlesInputUpdate
}

input AuthorInput @oneOf {
  create: ArticlesInputCreate
  update: ArticlesInputUpdate
}

// With usage:

mutation {
  articles(
    input: {
      create: {
        title: "new article"
        author: { update: { id: 5, name: "Rijk" } }
      }
    }
  )
}
```

but that's already a lot nicer (IMO!) than needing separate `create_articles`

**Comment by GustavoBlaze** [comment]
Hello! Are there any updates on this?

I've been trying to create an application using GraphQL but this bug is being quite of a blocker. If there's no solution I'll end up changing the entire app to use REST or SDK instead.
